src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -196,11 +196,16 @@ void G1BarrierSetC2::pre_barrier(GraphKit* kit,
     assert(pre_val == nullptr, "loaded already?");
     assert(val_type != nullptr, "need a type");
 
+    if (use_ReduceInitialCardMarks() && obj == kit->just_allocated_object(kit->control())) {
```

Comment:
This isn't correct.  It doesn't ensure there aren't any stores to the same
location between the allocation and this store that might need to be tracked.
g1_can_remove_pre_barrier (below) does that.

Reply:
We've had lots of problems with safepoints sneaking into unexpected places.
That's why (Gen)ZGC uses late barrier expansion.  That's also one of the
motivations for https://openjdk.org/jeps/475.

Reply:
> This isn't correct. It doesn't ensure there aren't any stores to the same location between the allocation and this store that might need to be tracked. g1_can_remove_pre_barrier (below) does that.

Hi Kim, g1_can_remove_pre_barrier is not sufficient which only eliminates stores with previous "null" value from object initialization. But actually no matter the original value is null or not stores to new allocated objects don't need SATB barrier at all. SATB pre-write barriers are delete protection barrier to protect reference disconnected from graph which is not necessary for new allocated objects. Any stored pointers into new alllocated objects are guaranted not white. GenZGC uses the same way the store barrier could be removed only if allocations dominate without safepoint blocking. The pre-write barrier elimination should be same to post-write barrier for new allocated objects.

Reply:
> We've had lots of problems with safepoints sneaking into unexpected places. That's why (Gen)ZGC uses late barrier expansion. That's also one of the motivations for https://openjdk.org/jeps/475.

That's problems with Load barrier because Load nodes in C2 could be rescheduled into other basic blocks which would cause safepoint between Load barrier and Load. Store nodes don't have this problem. This PR doesn't conflict with JEP475 and would reduce the barrier data in very early stage.

Reply:
I think that store capturing of initializing stores already removes most of the barriers of this category. We do that a bit later on. We find initializing stores onto newly allocated objects, and replace the store with barriers, with a store without barriers. That one usually elides a large portion of store barriers. Did you find any example where you have a newly allocated object with stores that are not initializing, hence not elided, which unnecessarily invoked a pre-write barrier, but not a post-write barrier? Just trying to find out what the problem space is, that this fixes.

Reply:
> I think that store capturing of initializing stores already removes most of the barriers of this category. We do that a bit later on. We find initializing stores onto newly allocated objects, and replace the store with barriers, with a store without barriers. That one usually elides a large portion of store barriers. Did you find any example where you have a newly allocated object with stores that are not initializing, hence not elided, which unnecessarily invoked a pre-write barrier, but not a post-write barrier? Just trying to find out what the problem space is, that this fixes.

Hi Erik,  I found examples not filtered by g1_can_remove_pre_barrier in testing. But I just did some statistics on SPECjbb2015 that if g1_can_remove_pre_barrier ran first it would elide most of the pre-barriers and "obj == kit->just_allocated_object"  only found very few remaining opportunities. If we run condition "obj == kit->just_allocated_object" first, it would cover ~30% opportunities. I think technically this PR should be correct but it's up to reviewers to decide if we practically need it.

Reply:
Did you check how many of the stores where g1_can_remove_pre_barrier said false and you would have said true, were elided anyway during store capturing (cf. InitializeNode::capture_store), or as part of G1BarrierSetC2::eliminate_gc_barrier? In other words, how many barriers are you eliding, that were not in fact already elided, just a bit later on?

Reply:
> Did you check how many of the stores where g1_can_remove_pre_barrier said false and you would have said true, were elided anyway during store capturing (cf. InitializeNode::capture_store), or as part of G1BarrierSetC2::eliminate_gc_barrier? In other words, how many barriers are you eliding, that were not in fact already elided, just a bit later on?

There is only 1 store that g1_can_remove_pre_barrier return false and was elided by this PR in JBB.

Reply:
> > Did you check how many of the stores where g1_can_remove_pre_barrier said false and you would have said true, were elided anyway during store capturing (cf. InitializeNode::capture_store), or as part of G1BarrierSetC2::eliminate_gc_barrier? In other words, how many barriers are you eliding, that were not in fact already elided, just a bit later on?
> 
> 
> 
> There is only 1 store that g1_can_remove_pre_barrier return false and was elided by this PR in JBB.

Okay. That's what I expected. Given that we are about to remove all of this code in favour of more robust late barrier expansion, I feel like we can live without that one extra store barrier for now.

Reply:
> > > Did you check how many of the stores where g1_can_remove_pre_barrier said false and you would have said true, were elided anyway during store capturing (cf. InitializeNode::capture_store), or as part of G1BarrierSetC2::eliminate_gc_barrier? In other words, how many barriers are you eliding, that were not in fact already elided, just a bit later on?
> > 
> > 
> > There is only 1 store that g1_can_remove_pre_barrier return false and was elided by this PR in JBB.
> 
> Okay. That's what I expected. Given that we are about to remove all of this code in favour of more robust late barrier expansion, I feel like we can live without that one extra store barrier for now.

ok. That's fairly reasonable.

Reply:
> > > 
> > > There is only 1 store that g1_can_remove_pre_barrier return false and was elided by this PR in JBB.
> > 
> > 
> > Okay. That's what I expected. Given that we are about to remove all of this code in favour of more robust late barrier expansion, I feel like we can live without that one extra store barrier for now.
> 
> ok. That's fairly reasonable.

Could you please withdraw this PR so that it does not show up in the list of open PRs?

Thanks,
  Thomas